### PR TITLE
CA-400560: Support tilde in RPM version/release comparison

### PR DESCRIPTION
**Pull request by changlei-li**
1. Fix version segment division error.
2. Support tilde in RPM version/release comparison.

### CODE REVIEW

File: ocaml/xapi/rpm.ml
Comment (minglumlu) [inline ocaml/xapi/rpm.ml:247]: A general approach of parsing is preferred than the specific functions.

For example of "1.2.3~rc1~beta" "<" "1.2.3~rc1", the `segment` lists would be:
`[Int 1; Int 2; Int 3; Tilde; Str "rc"; Int 1; Tilde; Str "beta"]` and `[Int 1; Int 2; Int 3; Tilde; Str "rc"; Int 1]`. It would be easy to update the pattern matching in `compare_segments`.

Comment (changlei-li) [reply ocaml/xapi/rpm.ml:247]: Accept

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [inline ocaml/xapi/rpm.ml:176]: I find this is incorrect. It should be 1.2.3 < 1.2.3a
```bash
$ rpmdev-vercmp 1.2.3 1.2.3a
1.2.3 < 1.2.3a
```

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [reply ocaml/xapi/rpm.ml:245]: Agree. Thanks for pointing out.

File: ocaml/xapi/rpm.ml
Comment (lindig) [reply ocaml/xapi/rpm.ml:247]: I support the proposal to parse a string into tagged components and to define order over these lists.

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [reply ocaml/xapi/rpm.ml]: I accept Ming's suggestion to add parse ~ in segment. So this is function is not used anymore.

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [reply ocaml/xapi/rpm.ml:176]: Fixed.

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [reply ocaml/xapi/rpm.ml]: According to the comment, I split the version string to segment of Int, Str and Tilde and fix some errors. I split the commits to fix segment division and support tilde. Because there is a big difference from the initial version, I push -f directly for not affecting the review experience.

Comment (lindig) [reply ocaml/xapi/rpm.ml]: I don't know if the tilde could be considered just a character in a `Str` and you would get the desired order.

Comment (changlei-li) [reply ocaml/xapi/rpm.ml]: @lindig, in your example, bash command `sort` thinks `~abc` < `abc` because the locale settings.  For example in the en_US.UTF-8 locale, the tilde character is often treated as less than alphanumeric characters.
In `man sort`
> *** WARNING *** The locale specified by the environment affects sort order.  Set LC_ALL=C to get the traditional sort order that uses native byte values.

And in OCaml, String.compare uses lexicographical order.
``` 
utop # List.sort String.compare ["ab"; "abc"; "~abc"];;
- : string list = ["ab"; "abc"; "~abc"]
```

Tilde is more complicated than just a character in a `Str`.  It may be along with letters or numbers, and even used separately. You can find more examples in the ut.
Generally, tilde (and the part after it) in the version string stands for pre-release:
1. Compare the formal version part first then compare the pre-release version part.
2. Version with tilde is earlier than the same version without tilde.

Comment (lindig) [reply ocaml/xapi/rpm.ml]: It the end, a total order needs to be defined between two arbitrary versions. Hence, we can't rely on "same versions without a tilde" because this is just a special case. So I would like to see this total order being defined over lists of `Str`, `Int` and `Tilde` or similar.

Comment (changlei-li) [reply ocaml/xapi/rpm.ml]: Of course. It's just a general statement to help comprehension. The function `compare_segments`  in the code define how to compare segment list.
Function  `compare_version_segment ` define how to compare single segment.

```ocaml
  type order = LT | EQ | GT

  type version_segment = Int of int | Str of string | Tilde

  let order_of_int = function 0 -> EQ | r when r > 0 -> GT | _ -> LT

  let rec compare_segments l1 l2 =
    match (l1, l2) with
    | c1 :: t1, c2 :: t2 -> (
      match compare_version_segment c1 c2 with
      | EQ ->
          compare_segments t1 t2
      | r ->
          r
    )
    | Tilde :: _, [] ->
        LT
    | [], Tilde :: _ ->
        GT
    | _ :: _, [] ->
        GT
    | [], _ :: _ ->
        LT
    | [], [] ->
        EQ

  let compare_version_segment s1 s2 =
    match (s1, s2) with
    | Int i1, Int i2 ->
        Int.compare i1 i2 |> order_of_int
    | Str s1, Str s2 ->
        String.compare s1 s2 |> order_of_int
    | Tilde, Tilde ->
        EQ
    | Int _, Str _ ->
        GT
    | Str _, Int _ ->
        LT
    | Tilde, _ ->
        LT
    | _, Tilde ->
        GT
```
I think the code is clear to specify the order over segment lists.

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [reply ocaml/xapi/rpm.ml:181]: Accept. I will push a new commit to modify it. Maybe before merge, I need to squash some commits to make them clear.

File: ocaml/xapi/rpm.ml
Comment (lindig) [reply ocaml/xapi/rpm.ml]: The code is clear; this needs to be tested against the examples to show it has the desired order.

File: ocaml/xapi/rpm.ml
Comment (lindig) [inline ocaml/xapi/rpm.ml:180]: Do we have to detect when a version string contains illegal characters? For example, the above regex does not allow a underscore to be part of the string.

File: ocaml/xapi/rpm.ml
Comment (lindig) [inline ocaml/xapi/rpm.ml:188]: Be aware that `int_of_string "0xabc"` does not throw an exception as you might expect.,

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [reply ocaml/xapi/rpm.ml]: The description is just like code :)
(1) Compare two segments:
For different kinds of segments:
Int > Str > Tilde
If they are the same kind, just compare the value (string or int)

(2) Compare two segment lists:
Compare segments in the two lists in order as (1).
If the order is determined (Not Eq)  -> end.
If they are equal, compare the next segments from the lists recursively.
If one list is empty, then Str/Int > Empty > Tilde -> end

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [reply ocaml/xapi/rpm.ml:180]: All illegal characters expect for `[alphabetic]+, [digit]+ and ~` are acted as delimiter:
```
1.2#3_45abc -> ["1"; "2"; "3"; "45"; "abc"]
1.2.3α45 -> ["1"; "2"; "3"; "45"]
12汉字45 -> ["12"; "45"]
```

File: ocaml/xapi/rpm.ml
Comment (changlei-li) [reply ocaml/xapi/rpm.ml:188]: Yes. No need to test the string is all digits when convert string to Int as the original implementation, because the split method has promised the string is all digits or all alphabetic. There is no situation like "0b11" may mislead `int_of_string`. So just use `int_of_string` exception to  distinguish Str and Int.
Due to the function has precondition that the input is from the split result, it is not a complete function to convert `string` to `version_segment`. So I just put it in the `normalize` function scope.

**Comment by changlei-li** [comment]
Hi maintainer, I need clean up my batches before merge.

File: ocaml/xapi/rpm.ml
Comment (lindig) [reply ocaml/xapi/rpm.ml:188]: An alternative would be to use `String.sscanf "%d"`. But you are correct that the problem does not arise because the split makes sure this is only digtits or only letters.

**Review by lindig — APPROVED** [review]

**Pull request merged**
